modules/flann/include/opencv2/flann/heap.h

```diff
@@ -49,117 +61,174 @@ namespace cvflann
-        if (count == length) {
+        if (size() == capacity()) {
             return;
         }
 
         heap.push_back(value);
-        static CompareT compareT;
-        std::push_heap(heap.begin(), heap.end(), compareT);
-        ++count;
+        std::push_heap(heap.begin(), heap.end(), greater<T>());
     }
 
-
-
     /**
-     * Returns the node of minimum value from the heap (top of the heap).
+     * \brief Returns the node of minimum value from the heap (top of the heap).
      *
-     * Params:
-     *     value = out parameter used to return the min element
-     * Returns: false if heap empty
+     * \param[out] value parameter used to return the min element
+     * \returns false if heap empty
      */
     bool popMin(T& value)
     {
-        if (count == 0) {
+        if (empty()) {
             return false;
         }
 
         value = heap[0];
-        static CompareT compareT;
-        std::pop_heap(heap.begin(), heap.end(), compareT);
+        std::pop_heap(heap.begin(), heap.end(), greater<T>());
         heap.pop_back();
-        --count;
 
         return true;  /* Return old last node. */
     }
+
+    /**
+     * \brief Returns a shared heap for the given memory pool ID.
+     *
+     * It constructs the heap if it does not already exists.
+     *
+     * \param poolId a user-chosen hashable ID for identifying the heap.
+     *     For thread-safe operations, using current thread ID is a good choice.
+     * \param capacity heap maximum capacity
+     * \param iterThreshold remove heaps that were not reused for more than specified iterations count
+     * \returns pointer to the heap
+     */
+    template <typename HashableT>
+    static cv::Ptr<Heap<T>> getPooledInstance(
+        const HashableT& poolId, const int capacity, const int iterThreshold = 2 * cv::getNumThreads())
+    {
+        static cv::Mutex mutex;
+        const cv::AutoLock lock(mutex);
+
+        struct HeapMapValueType {
+            cv::Ptr<Heap<T>> heapPtr;
+            int iterCounter;
+        };
+        typedef std::unordered_map<HashableT, HeapMapValueType> HeapMapType;
+
+        static HeapMapType heapsPool;
+        typename HeapMapType::iterator heapIt = heapsPool.find(poolId);
+
+        if (heapIt == heapsPool.end())
+        {
+            // Construct the heap as it does not already exists
+            HeapMapValueType heapAndTimePair = {cv::makePtr<Heap<T>>(capacity), 0};
+            const std::pair<typename HeapMapType::iterator, bool>& emplaceResult = heapsPool.emplace(poolId, std::move(heapAndTimePair));
+            CV_CheckEQ(emplaceResult.second, true, "Failed to insert the heap into its memory pool");
+            heapIt = emplaceResult.first;
+        }
+        else
+        {
+            CV_CheckEQ(heapIt->second.heapPtr.use_count(), 1, "Cannot modify a heap that is currently accessed by another caller");
+            heapIt->second.heapPtr->clear();
+            heapIt->second.heapPtr->reserve(capacity);
```

Comment:
It makes sense to store smart pointer `heapIt->second.heapPtr` in local variable (to be returned as result) before applying quite dangerous "cleanup" procedure with `.erase()` calls which may invalidate `heapIt` iterator.

Reply:
I added `CV_Assert(cleanupIt != heapIt)` before `.erase()` call, although it is actually impossible to assert, since `iterCounter=0`, and `iterThreshold>=2`.